### PR TITLE
[Bug Fix] - Fixed TokenListener crash on devices 8 and higher

### PR DESCRIPTION
**Pull request by Rcureton**
## What
- We updated the `FirebaseInstanceIdService` to `FirebaseMessagingService`
- We added `Firebase-job-dispatcher` to the project.
- We Removed `IntentService` and replaced it with `JobService`.

## Why
- We added `Firebase-job-dispatcher` so that on devices on OS 8 and higher we can run jobs in the background. It also supports back to API level 14.

- We upgraded the `FirebaseMessaging` import to use the new `FirebaseMessagingService` class which replaced `FirebaseInstanceIdService`. We also had to change the `intent-filter` in the manifest. 
    > A service that extends FirebaseMessagingService. This is required if you want to do any message handling beyond receiving notifications on apps in the background. To receive notifications in foregrounded apps, to receive data payload, to send upstream messages, and so on, you must extend this service.

### CODE REVIEW

File: app/src/main/java/com/kickstarter/libs/DeviceRegistrar.java [hunk 1]
```diff
@@ -29,7 +31,13 @@ public void registerDevice() {
       return;
     }
 
-    this.context.startService(new Intent(this.context, RegisterService.class));
+    final FirebaseJobDispatcher dispatcher = new FirebaseJobDispatcher(new GooglePlayDriver(this.context));
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/libs/DeviceRegistrar.java]: Looks like we're doing the same thing 3 times, which seems like a Util class would be helpful! Let's make a Util class for Jobs™. So you'd pass in a context, the class, and the tag. And then we could make these tags constants!

Comment (eoji) [inline app/src/main/java/com/kickstarter/libs/DeviceRegistrar.java]: I'm also wondering if it's expense to keep making new `FirebaseJobDispatcher` objects and we should make it a singleton instead.

File: app/src/main/java/com/kickstarter/services/firebase/RegisterService.java [hunk 2]
```diff
@@ -18,35 +17,32 @@
...
     super.onCreate();
     ((KSApplication) getApplicationContext()).component().inject(this);
   }
 
+
   @Override
-  protected void onHandleIntent(final @Nullable Intent intent) {
-    Timber.d("onHandleIntent");
+  public boolean onStartJob(final JobParameters job) {
+    FirebaseInstanceId.getInstance().getInstanceId().addOnSuccessListener(instanceIdResult -> {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/services/firebase/RegisterService.java]: Nice job getting rid of the try catch

File: app/src/main/java/com/kickstarter/services/firebase/RegisterService.java [hunk 3]
```diff
@@ -18,35 +17,32 @@
...
-      if(ObjectUtils.isNotNull(token)) {
-        sendTokenToApi(token);
-        subscribeToGlobalTopic();
-      }
-    } catch (final Exception e) {
-      Timber.e("Failed to complete token refresh: %s", e);
-    }
+  @Override
+  public boolean onStopJob(final JobParameters job) {
+    return false;
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/services/firebase/RegisterService.java]: what does this mean and why do we return false?

Comment (Rcureton) [reply app/src/main/java/com/kickstarter/services/firebase/RegisterService.java]: I guess if we needed to kill the job we would make sure we dispose of it here.

File: app/src/main/java/com/kickstarter/libs/DeviceRegistrar.java (hunk 1 above)
Comment (Rcureton) [reply app/src/main/java/com/kickstarter/libs/DeviceRegistrar.java]: I'll make a file-level function for it!

File: app/src/main/java/com/kickstarter/services/firebase/RegisterService.java [hunk 4]
```diff
@@ -18,35 +17,32 @@
...
+      final String newToken = instanceIdResult.getToken();
+      Timber.d("newToken", newToken);
+      sendTokenToApi(newToken);
+      subscribeToGlobalTopic();
+    });
 
-    try {
-      final String token = FirebaseInstanceId.getInstance().getToken();
-      Timber.d("Token: %s", token);
+    return true;
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/services/firebase/RegisterService.java]: >Return `true` from this method if your job needs to continue running.`

I don't think that's what we want. When do we ever stop the job?

File: app/src/main/java/com/kickstarter/ApplicationModule.java [hunk 5]
```diff
@@ -177,6 +179,13 @@ static ApolloClientType provideApolloClientType(final @NonNull ApolloClient apol
     return new KSApolloClient(apolloClient);
   }
 
+  @Provides
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ApplicationModule.java]: Let's get rid of this.

Comment (Rcureton) [reply app/src/main/java/com/kickstarter/ApplicationModule.java]: I removed it in a previous PR

**Comment by eoji** [comment]
[[P2] TokenListenerService service crash on Oreo and above](https://trello.com/c/hPwUQhEw/946-p2-tokenlistenerservice-service-crash-on-oreo-and-above)

**Pull request merged**